setup/base/base.go

```diff
@@ -183,12 +183,21 @@ func NewBaseDendrite(cfg *config.Dendrite, componentName string, options ...Base
 			},
 		},
 	}
-	client := http.Client{Timeout: HTTPClientTimeout}
+	transport := *http.DefaultTransport.(*http.Transport)
+	if cfg.Global.HTTPClientTLSInsecureSkipVerify {
+		transport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
 	if cfg.FederationAPI.Proxy.Enabled {
-		client.Transport = &http.Transport{Proxy: http.ProxyURL(&url.URL{
+		transport.Proxy = http.ProxyURL(&url.URL{
 			Scheme: cfg.FederationAPI.Proxy.Protocol,
 			Host:   fmt.Sprintf("%s:%d", cfg.FederationAPI.Proxy.Host, cfg.FederationAPI.Proxy.Port),
-		})}
+		})
+	}
+	client := http.Client{
```

Comment:
I’m not *entirely* sure why we have this HTTP client at any rate — I think anything that previously used it is probably now using `apiHttpClient` instead. Is it really used?

Also this will probably only really do what you expect in monolith mode, ideally the component that is producing the HTTP client calls (I guess in your case, the push server) should be instantiating a HTTP client for the purpose. That said, having a function that returns a suitable HTTP client based on the config (proxy, TLS validation etc) would probably be a useful way to simplify this as well as in other places like the appservice API.

Reply:
You're right. It was replaced with `apiHttpClient` 4d ago: https://github.com/matrix-org/dendrite/commit/ec716793eb86e7723ecfaa356e148334bc3921eb#diff-794b74a8e5bab56b4fdac37a8aeb4fbc1d6eaa5218341d75761abad8af1d9729L269. Federation was the only case I know of using it.

So I'm guessing that means `client` should be removed now? (I thought it was neat there was a separate client for external requests, though. You have far less control (and probably higher security requirements) over those requests. So made sense to adjust timeouts and security params separately.)

BTW, the reason I made this a separate PR is because I need to modify Sytest to enable this option before I can post the Pushgateway PR. Here's the end-goal: https://github.com/tommie/dendrite/blob/serving/setup/base.go#L308

> Also this will probably only really do what you expect in monolith mode

Even in polyliths, the base is still shared. Why is it not doing what I'd expect? Calling code here: https://github.com/tommie/dendrite/blob/serving/cmd/dendrite-polylith-multi/personalities/pushserver.go#L25

I don't understand your preferred approach. Even now, the `apiHttpClient` is injected into e.g. Federation API. Are you suggesting doing something different? (It sounds like you want to inject a client factory; an indirection I don't understand the purpose of.)

Reply:
> So I'm guessing that means client should be removed now? 

Well, if there's a good reason for using a shared client like this then I guess we can keep it, but it isn't really clear that a shared client is any better than components just building a `http.Client` with timeout and proxy values that make sense for that specific workload. It might just be clearer to remove it.

Note that `apiHttpClient` is specifically different because it forces H2C instead of plain HTTP, which we use when talking *between* components in polylith mode, but we strictly do not use it to make external HTTP calls.

> BTW, the reason I made this a separate PR is because I need to modify Sytest to enable this option before I can post the Pushgateway PR. Here's the end-goal: https://github.com/tommie/dendrite/blob/serving/setup/base.go#L308

This feels misplaced. The various `FooHTTPClient()` functions here are for returning shaped clients for talking to *internal APIs* so that components can communicate with each other. Surely the only thing talking to push gateways directly is this new push server? In which case why would the `base` package care about the client used?

> Even in polyliths, the base is still shared

Yes, you're right, `NewBaseDendrite` gets called before the polylith personalities are attached.

> I don't understand your preferred approach. Even now, the apiHttpClient is injected into e.g. Federation API. 

I don't think this is true (at least it shouldn't be). The `apiHttpClient` isn't really injected anywhere — it's used to return a HTTP-powered API implementation for a given component's API surface. The federation API makes its own `http.Client` indirectly via the `gomatrixserverlib.FederationClient`, the media API uses a `gomatrixserverlib.Client` etc.

Reply:
Ah, thanks for the explanation. Makes sense.

In that case, I think my we/I have three options:

1. Create the global config option here, enable it in Sytest, then submit the Push gateway code with Sytest whitelisting.
1. Create the pushserver-specific config option here, enable it in Sytest, then submit the Push gateway code with Sytest whitelisting.
2. Create the Push gateway code with a pushgateway-specific config option to disable TLS verification, enable it in Sytest, then submit Sytest whitelisting.

Since I'd love for the big PR to pass Sytests rather than do that as a follow-up, I'd prefer the (1) or (2) flows. Given that we won't have a global HTTP client, (2) makes the most sense.

Updated code pushed.